module/zfs/zil.c

```diff
@@ -723,9 +725,39 @@ zil_create(zilog_t *zilog)
 	 * (zh is part of the MOS, so we cannot modify it in open context.)
 	 */
 	if (tx != NULL) {
+		/*
+		 * If "zilsaxattr" feature is enabled on zpool, then activate
+		 * it.
+		 */
+		if (spa_feature_is_enabled(zilog->zl_spa,
+		    SPA_FEATURE_ZILSAXATTR)) {
+			mutex_enter(&ds->ds_lock);
+			ds->ds_feature_activation[SPA_FEATURE_ZILSAXATTR] =
+			    (void *)B_TRUE;
+			mutex_exit(&ds->ds_lock);
+		}
 		dmu_tx_commit(tx);
 		txg_wait_synced(zilog->zl_dmu_pool, txg);
+	} else {
+		if (spa_feature_is_enabled(zilog->zl_spa,
+		    SPA_FEATURE_ZILSAXATTR) &&
+		    !dsl_dataset_feature_is_active(ds,
+		    SPA_FEATURE_ZILSAXATTR)) {
+			tx = dmu_tx_create(zilog->zl_os);
```

Comment:
Notes from conversation earlier today: this branch covers the case where we enable the feature on a zpool that has existing ZIL headers.

Reply:
yes